lib/api/model/widget.dart

```diff
@@ -0,0 +1,131 @@
+import '../../log.dart';
+import '../../model/store.dart';
+import 'submessage.dart';
+
+/// States of a poll Zulip widget.
+///
+/// See also:
+/// - https://zulip.com/help/create-a-poll
+/// - https://github.com/zulip/zulip/blob/304d948416465c1a085122af5d752f03d6797003/web/shared/src/poll_data.ts
+class Poll {
+  Poll({
+    required this.submessages,
+    required this.pollSenderId,
+    required this.question,
+    required final List<String> options,
+  }) {
+    for (int index = 0; index < options.length; index += 1) {
+      // [null] is necessary for creating initial poll options.
+      _addOption(null, options[index], optionIndex: index);
+    }
+  }
+
+  factory Poll.fromSubmessages(
+    List<Submessage> submessages,
+    {required int senderId, required PollWidgetData widgetData}
+  ) {
+    assert(submessages.isNotEmpty);
+    final widgetEventSubmessages = submessages.sublist(1);
+
+    final poll = Poll(
+      submessages: submessages,
+      pollSenderId: senderId,
+      question: widgetData.extraData.question,
+      options: widgetData.extraData.options,
+    );
+
+    for (final submessage in widgetEventSubmessages) {
+      final event = PollEvent.fromJson(submessage.content as Map<String, Object?>);
+      poll.applyEvent(submessage.senderId, event);
+    }
+    return poll;
+  }
+
+  /// See [Message._pollToJson].
+  final List<Submessage> submessages;
```

Comment:
This exists purely for testing as mentioned in the commit message, specifically, for `prepareMessages` calls, where we serialize the `GetMessageResult` objects containing message objects with submessages. I think a way to get around this is refactoring out the part where `Message` objects are used in the test helpers, and modify the mappings directly, then returning empty lists should be fine.

Reply:
I see. Yeah, that solution seems fine.

I think it would be a solvable problem to make `Message.toJson` produce an accurate `submessages`. But it would be kind of a normalized version — e.g. I think if someone voted, unvoted, then voted again, it'd record only the final vote because that's all the `Poll` object remembers. So for some tests that might not be adequate anyway.